api/replica.go

```diff
@@ -14,20 +19,66 @@ type Replica struct {
 	handler   http.Handler
 	tlsConfig *tls.Config
 	primary   string
+	addr      string // the address of this replica
+	// mu is the mutex used for locking the primary.
+	mu sync.RWMutex
+	// pWait is used to block execution of ServeHTTP until the primary is ready
+	// primaryWait is only used by readers, and is populated by the output of
+	// mu.RLocker, which means the lock associated with it is actually a read
+	// lock.
+	primaryWait *sync.Cond
+
+	// capture the hijack function for easier unit testing
+	hijack func(*tls.Config, string, http.ResponseWriter, *http.Request) error
 }
 
 // NewReplica creates a new API replica.
-func NewReplica(handler http.Handler, tlsConfig *tls.Config) *Replica {
-	return &Replica{
+func NewReplica(handler http.Handler, tlsConfig *tls.Config, addr string) *Replica {
+	r := &Replica{
 		handler:   handler,
 		tlsConfig: tlsConfig,
+		addr:      addr,
+		hijack:    hijack,
 	}
+	r.primaryWait = sync.NewCond(r.mu.RLocker())
```

Comment:
nit: can this be done inline in the Replica{} init above ?

Reply:
No, because we need the replica's mutex to the the `RLocker`.